allow decals to be rotated

microlith57 opened the pull request:
allows a `rotation` float (in radians) to be added to decals in the map data, and rotates decals accordingly.

closes #467

microlith57 committed:
initial decal rotation implementation

microlith57 commented on Celeste.Mod.mm/Patches/Decal.cs:
```diff
@@ -45,22 +48,43 @@ public patch_Decal(string texture, Vector2 position, Vector2 scale, int depth)
             }
         }
 
-        [MonoModIgnore]
```
maybe `patch_DecalImage` should be above `patch_CoreSwapImage`?

microlith57 commented on Celeste.Mod.mm/Patches/Level.cs:
```diff
@@ -518,6 +519,12 @@ namespace MonoMod {
     class PatchLevelLoaderAttribute : Attribute { }
 
+    /// <summary>
+    /// Patch the Godzilla-sized level loading method instead of reimplementing it in Everest.
```
should have a better comment

bigkahuna443 replied:
would prefer a more useful comment here

microlith57 committed:
make decal rotation also rotate mirror masks

microlith57 committed:
also rotate banner decals, scared decals

bigkahuna443 commented on Celeste.Mod.mm/Patches/Decal.cs:
```diff
@@ -45,22 +48,79 @@ public patch_Decal(string texture, Vector2 position, Vector2 scale, int depth)
+                                 origin: new Vector2(tex.Width / 2, tex.Height / 2 - i * SliceSize),
+                                 Color.White,
+                                 decal.scale,
+                                 decal.Rotation);
```
I think this syntax of one line per item makes sense with something like LINQ where each line is a different operation, but it looks a bit odd here. Might try condensing it to e.g. have the "origin" field on line 78 and the rest of the fields on line 79. applies elsewhere too

bigkahuna443 commented on Celeste.Mod.mm/Patches/Decal.cs:
```diff
@@ -45,22 +48,79 @@ public patch_Decal(string texture, Vector2 position, Vector2 scale, int depth)
+            public List<List<MTexture>> Segments;
+#pragma warning restore CS0649
+
+            private patch_Decal decal => (patch_Decal) base.Entity;
```
This property is public in the original class, so this patch causes a new property to be created

bigkahuna443 commented on Celeste.Mod.mm/Patches/Decal.cs:
```diff
@@ -45,22 +48,79 @@ public patch_Decal(string texture, Vector2 position, Vector2 scale, int depth)
+            }
+
+            public extern void orig_Render();
+            public override void Render() {
```
you're not using the orig method for anything so you can just `[MonoModReplace]` this instead

bigkahuna443 commented on Celeste.Mod.mm/Patches/Decal.cs:
```diff
@@ -45,22 +48,79 @@ public patch_Decal(string texture, Vector2 position, Vector2 scale, int depth)
+
+            public extern void orig_Render();
+            public override void Render() {
+                ((patch_Decal) Entity).textures[(int) ((patch_Decal) Entity).frame].DrawCentered(
```
the component already has a `Decal` field to get the decal which you can use instead of casting `Entity`, makes IL a bit cleaner

bigkahuna443 commented on Celeste.Mod.mm/Patches/Decal.cs:
```diff
@@ -82,7 +142,11 @@ private class FlagSwapImage : Component {
             public override void Render() {
                 if (activeTextures.Count > 0)
-                    activeTextures[(int) frame % activeTextures.Count].DrawCentered(Decal.Position, Color.White, Decal.scale);
+                    activeTextures[(int) frame % activeTextures.Count].DrawCentered(
```
yeah there are longer lines than this already, you could just have it be on one lol

bigkahuna443 commented on Celeste.Mod.mm/Patches/Decal.cs:
```diff
@@ -45,22 +48,79 @@ public patch_Decal(string texture, Vector2 position, Vector2 scale, int depth)
-        private class CoreSwapImage : Component {
-            public CoreSwapImage(MTexture hot, MTexture cold) : base(active: false, visible: true) {
-                // no-op
+        private class patch_Banner : Component {
```
Overall comment: it's a little unfortunate to have to add all these replaced functions. Looking at the IL it looks like you could get away with some shared IL patches. Just move before the `Draw` or `DrawCentered` call -> load the `Rotation` field -> call the function that takes in the rotation -> pop the old one. Would be a lot cleaner if it works.

(left some other comments in case you keep the original implementation though)

bigkahuna443 commented on Celeste.Mod.mm/Patches/Decal.cs:
```diff
@@ -45,22 +48,79 @@ public patch_Decal(string texture, Vector2 position, Vector2 scale, int depth)
+                MTexture tex = decal.textures[(int) decal.frame];
+                List<MTexture> segments = Segments[(int) decal.frame];
+                for (int i = 0; i < segments.Count; i++) {
+                    float num = (EaseDown ? ((float) i / (float) segments.Count) : (1f - (float) i / (float) segments.Count)) * WindMultiplier;
```
there are some unnecessary casts here from the decompilation that my IDE catches

bigkahuna443 commented on Celeste.Mod.mm/Patches/DecalData.cs:
```diff
@@ -0,0 +1,9 @@
+namespace Celeste {
+    class patch_DecalData : DecalData {
+
+#pragma warning disable CS0649 // The field is never assigned and will always be null
```
Not a huge deal but I think these are cleaner as a global disable at the top of the file

bigkahuna443 commented on Celeste.Mod.mm/Patches/Level.cs:
```diff
@@ -37,7 +37,7 @@ class patch_Level : Level {
 
         public delegate Entity EntityLoader(Level level, LevelData levelData, Vector2 offset, EntityData entityData);
         public static readonly Dictionary<string, EntityLoader> EntityLoaders = new Dictionary<string, EntityLoader>();
-        
```
whitespace change 😱

microlith57 replied:
my ide does it automatically

bigkahuna443 commented on Celeste.Mod.mm/Patches/Decal.cs:
```diff
@@ -262,7 +388,6 @@ private class FlagSwapImage : Component {
     }
 
     public static class DecalExt {
-
```
I think it's actually supposed to be convention to have a blank line at the beginning of a class lol

bigkahuna443 commented on Celeste.Mod.mm/Patches/Level.cs:
```diff
@@ -186,7 +186,8 @@ class patch_Level : Level {
 
-        [PatchLevelLoader] // Manually manipulate the method via MonoModRules
+        [PatchLevelLoaderDecalCreation] // Manually manipulate the method via MonoModRules
+        [PatchLevelLoader]
```
I took a look and only found one other usage like this in Everest... It's definitely easier to read patches when they're split up, but can also create issues if they're order-dependent or one changes IL the other depends on. Doesn't seem like there's a conflict here though and I agree that it's out of scope, we should probably try to combine them if/when we refactor the old patches though

EDIT: nitpicky but if you put the new attribute below the old one it'd be a 1-line patch instead of 3 smh

bigkahuna443 commented on Celeste.Mod.mm/Patches/Level.cs:
```diff
@@ -639,6 +646,45 @@ static partial class MonoModRules {
+            while (cursor.TryGotoNext(MoveType.Before,
+                                      // we are looking for one of the possible decal depths (fg or bg)...
+                                      instr => instr.MatchLdcI4(-10500)
+                                            || instr.MatchLdcI4(9000),
```
you can use `Celeste.Depths.FGDecals` and `Celeste.Depths.BGDecals` for these. could make more sense to add a new Decal constructor though and swap it out instead of adding it to the existing reference?

microlith57 replied:
i tried that but it's a massive pain to get a reference to one of several constructors (they're all called `.ctor`); if you know a way to do that i'll implement it

Kalobi replied:
`GetConstructor(Type[])`?

bigkahuna443 commented on Celeste.Mod.mm/Patches/LevelData.cs:
```diff
@@ -63,5 +70,61 @@ we found
+                         stfld float32 Celeste.DecalData::Rotation
+                IL_0a7b: callvirt instance void class [mscorlib]System.Collections.Generic.List`1<class Celeste.DecalData>::Add(!0)
+
+                in both places where a DecalData instance is added to a List
```
I definitely think a C# comment would be more helpful here than IL, we can already see the IL below

bigkahuna443 commented on Celeste.Mod.mm/Patches/LevelData.cs:
```diff
@@ -63,5 +70,61 @@ we found
+                cursor.Emit(OpCodes.Stfld, f_DecalDataRotation);
+
+                // make sure we won't find this location the next time we TryGotoNext
+                cursor.GotoNext(MoveType.After, instr => instr.MatchCallvirt("System.Collections.Generic.List`1<Celeste.DecalData>", "Add"));
```
this is a bit confusing in combo with the while loop, it's the very next instruction so you can probably just increment the cursor

microlith57 committed:
better decal rotation implementation

microlith57 committed:
add a new Decal::.ctor that takes rotation; use it

microlith57 commented:
huh, apparently i can only request one review at a time

microlith57 committed:
set default decal rotation back to 0.0f

bigkahuna443 committed:
Add mirror mask patch + cleanup

Swapped all the replaced decal mirror mask functions with a general patch. Reworked other patches to be a bit more standard as well.

bigkahuna443 committed:
Fix missing default argument

microlith57 committed:
further cleanup
